utils/signature/verify.go

```diff
@@ -7,58 +7,112 @@ SPDX-License-Identifier: Apache-2.0
-	Scheme    Scheme
-	PublicKey PublicKey
+	thresholdVerifier ThresholdVerifier
+	signatureVerifier policies.Policy
+	Type              protoblocktx.PolicyType
+	Policy            Policy
 }
 
 // NewNsVerifier creates a new namespace verifier according to the implementation scheme.
-func NewNsVerifier(scheme Scheme, key PublicKey) (*NsVerifier, error) {
+func NewNsVerifier(p *protoblocktx.NamespacePolicy, idDeserializer msp.IdentityDeserializer) (*NsVerifier, error) {
 	res := &NsVerifier{
-		Scheme:    strings.ToUpper(scheme),
-		PublicKey: key,
+		Type:   p.Type,
+		Policy: p.Policy,
 	}
 	var err error
-	switch res.Scheme {
-	case NoScheme, "":
-		res.verifier = nil
-	case Ecdsa:
-		res.verifier, err = NewEcdsaVerifier(key)
-	case Bls:
-		res.verifier, err = NewBLSVerifier(key)
-	case Eddsa:
-		res.verifier = &EdDSAVerifier{PublicKey: key}
+
+	switch p.Type {
+	case protoblocktx.PolicyType_THRESHOLD_RULE:
+		policy := &protoblocktx.ThresholdRule{}
+		if merr := proto.Unmarshal(p.Policy, policy); merr != nil {
+			return nil, merr
+		}
+
+		switch policy.Scheme {
+		case NoScheme, "":
+			res.thresholdVerifier = nil
+		case Ecdsa:
+			res.thresholdVerifier, err = NewEcdsaVerifier(policy.PublicKey)
+		case Bls:
+			res.thresholdVerifier, err = NewBLSVerifier(policy.PublicKey)
+		case Eddsa:
+			res.thresholdVerifier = &EdDSAVerifier{PublicKey: policy.PublicKey}
+		default:
+			return nil, errors.Newf("scheme '%v' not supported", policy.Scheme)
+		}
+	case protoblocktx.PolicyType_SIGNATURE_RULE:
+		pp := cauthdsl.NewPolicyProvider(idDeserializer)
+		res.signatureVerifier, _, err = pp.NewPolicy(p.Policy)
 	default:
-		return nil, errors.Newf("scheme '%v' not supported", scheme)
+		return nil, errors.Newf("policy type '%v' not supported", p.Type)
 	}
 	return res, err
 }
 
 // VerifyNs verifies a transaction's namespace signature.
 func (v *NsVerifier) VerifyNs(txID string, tx *protoblocktx.Tx, nsIndex int) error {
-	if nsIndex < 0 || nsIndex >= len(tx.Namespaces) || nsIndex >= len(tx.Signatures) {
+	if nsIndex < 0 || nsIndex >= len(tx.Namespaces) || nsIndex >= len(tx.Endorsements) {
 		return errors.New("namespace index out of range")
 	}
-	if v.verifier == nil {
-		return nil
+
+	switch v.Type {
+	case protoblocktx.PolicyType_THRESHOLD_RULE:
+		if v.thresholdVerifier == nil {
+			return nil
+		}
+		digest, err := DigestTxNamespace(txID, tx.Namespaces[nsIndex])
+		if err != nil {
+			return err
+		}
+		return v.thresholdVerifier.Verify(digest, tx.Endorsements[nsIndex].EndorsementsWithIdentity[0].Endorsement)
+	case protoblocktx.PolicyType_SIGNATURE_RULE:
+		data, err := ASN1MarshalTxNamespace(txID, tx.Namespaces[nsIndex])
+		if err != nil {
+			return err
+		}
+		signedData := make([]*protoutil.SignedData, len(tx.Endorsements[nsIndex].EndorsementsWithIdentity))
+		for i, s := range tx.Endorsements[nsIndex].EndorsementsWithIdentity {
+			idBytes, err := msp.NewSerializedIdentity(s.Identity.MspId, s.Identity.GetCertificate())
```

Comment:
major: What about `CertificateId`?
If the MSP doesn't have support for this yet, then I suggest adding a TODO here.
Also, this code ignores the fact that `GetCertificate()` can return `nil`.